drv/i2c-devices/src/pca9956b.rs

```diff
@@ -0,0 +1,332 @@
+    ALLCALLADR = 0x3E,
+    PWMALL = 0x3F,
+    IREFALL = 0x40,
+    EFLAG0 = 0x41,
+    EFLAG1 = 0x42,
+    EFLAG2 = 0x43,
+    EFLAG3 = 0x44,
+    EFLAG4 = 0x45,
+    EFLAG5 = 0x46,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum LedErr {
+    NoError = 0b00,
+    ShortCircuit = 0b01,
+    OpenCircuit = 0b10,
+    Invalid = 0b11,
+}
+
+impl Default for LedErr {
+    fn default() -> Self {
+        LedErr::NoError
+    }
+}
+
+impl From<u8> for LedErr {
+    fn from(i: u8) -> Self {
+        match i {
+            0 => LedErr::NoError,
+            1 => LedErr::ShortCircuit,
+            2 => LedErr::OpenCircuit,
+            _ => LedErr::Invalid,
+        }
+    }
+}
+
+/// LedErrSummary is used to summarize the types of errors seen
+/// overtemp - true if MODE2_OVERTEMP was 1
+/// short_circuit - number of LEDs with LedErr::ShortCircuit
+/// open_circuit - number of LEDs with LedErr::OpenCircuit
+/// invald - number of LEDs with LedErr::Invalid
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+pub struct LedErrSummary {
+    pub overtemp: bool,
+    pub short_circuit: u8,
+    pub open_circuit: u8,
+    pub invalid: u8,
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+struct Pca9956BErrorState {
+    led_errors: [LedErr; NUM_LEDS],
+    overtemp: bool,
+}
+
+impl Pca9956BErrorState {
+    fn summary(&self) -> LedErrSummary {
+        let mut summary = LedErrSummary {
+            overtemp: self.overtemp,
+            ..Default::default()
+        };
+
+        for err in self.led_errors {
+            if err == LedErr::OpenCircuit {
+                summary.open_circuit += 1;
+            } else if err == LedErr::ShortCircuit {
+                summary.short_circuit += 1;
+            } else if err == LedErr::Invalid {
+                summary.invalid += 1;
+            }
+        }
+
+        summary
+    }
+}
+
+/// Auto-increment flag is Bit 7 of the control register. Bits 6..0 are address.
+const CTRL_AUTO_INCR_MASK: u8 = 1 << 7;
+/// The MODE2 OVERTEMP bit indicates if an overtempature condition has occurred
+const MODE2_OVERTEMP_MASK: u8 = 1 << 7;
+/// The MODE2 ERROR bit indicates if any error conditions are in EFLAGn
+const MODE2_ERROR_MASK: u8 = 1 << 6;
+/// The MODE2 CLRERR bit clears all error conditions in EFLAGn
+const MODE2_CLRERR_MASK: u8 = 1 << 4;
+/// The MODE2 reserved bits have a defined pattern (0b101) and are read only.
+/// They will be used to validate a PCA9956B with humility
+const MODE2_RSVD_MASK: u8 = 0x7;
+const MODE2_RSVD: u8 = 0x5;
+
+pub struct Pca9956B {
+    device: I2cDevice,
+}
+
+pub const NUM_LEDS: usize = 24;
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Error {
+    /// The low-level I2C communication returned an error
+    I2cError(ResponseCode),
+
+    /// The LED index is too large
+    InvalidLED(u8),
+}
+
+impl From<ResponseCode> for Error {
+    fn from(err: ResponseCode) -> Self {
+        Error::I2cError(err)
+    }
+}
+
+impl From<Error> for ResponseCode {
```

Comment:
I know that this pattern appears in a bunch of places, but we're trying to move away from it.

I'd recommend deleting this `impl From` and instead doing the `Error -> ResponseCode` conversion in `impl Validate`, e.g.
```diff
diff --git a/drv/i2c-devices/src/pca9956b.rs b/drv/i2c-devices/src/pca9956b.rs
index f9bda765..1c2bf743 100644
--- a/drv/i2c-devices/src/pca9956b.rs
+++ b/drv/i2c-devices/src/pca9956b.rs
@@ -186,15 +186,6 @@ impl From<ResponseCode> for Error {
     }
 }

-impl From<Error> for ResponseCode {
-    fn from(err: Error) -> Self {
-        match err {
-            Error::I2cError(code) => code,
-            _ => panic!(),
-        }
-    }
-}
-
 impl Pca9956B {
     pub fn new(device: &I2cDevice) -> Self {
         Self { device: *device }
@@ -323,9 +314,15 @@ impl Pca9956B {
 // which is the type of information we typically like to validate against.
 // MODE2[2:0] are set to read only an initialized to b101, so use that to
 // validate.
-impl Validate<Error> for Pca9956B {
-    fn validate(device: &I2cDevice) -> Result<bool, Error> {
-        let mode = Pca9956B::new(device).read_reg(Register::MODE2)?;
+impl Validate<ResponseCode> for Pca9956B {
+    fn validate(device: &I2cDevice) -> Result<bool, ResponseCode> {
+        let mode = Pca9956B::new(device).read_reg(Register::MODE2).map_err(
+            |e| match e {
+                // read_reg can only return Error::I2cError
+                Error::I2cError(e) => e,
+                _ => panic!(),
+            },
+        )?;

         Ok(mode & MODE2_RSVD_MASK == MODE2_RSVD)
     }
```

That way, the only possible `panic!` is in a well-known place; future users can't accidentally shoot themselves in the foot.